src/Proto3/Suite/DotProto/Internal.hs

```diff
@@ -151,24 +179,370 @@ data FindProtoResult
+isMap Map{} = True
+isMap _ = False
+
+--------------------------------------------------------------------------------
+--
+-- * Name resolution
+--
+
+concatDotProtoIdentifier :: MonadError CompileError m
+                         => DotProtoIdentifier -> DotProtoIdentifier -> m DotProtoIdentifier
+concatDotProtoIdentifier i1 i2 = case (i1, i2) of
+  (Qualified{}  ,  _           ) -> internalError "concatDotProtoIdentifier: Qualified"
+  (_            , Qualified{}  ) -> internalError "concatDotProtoIdentifier Qualified"
+  (Anonymous    , Anonymous    ) -> pure Anonymous
+  (Anonymous    , b            ) -> pure b
+  (a            , Anonymous    ) -> pure a
+  (Single a     , b            ) -> concatDotProtoIdentifier (Dots (Path [a])) b
+  (a            , Single b     ) -> concatDotProtoIdentifier a (Dots (Path [b]))
+  (Dots (Path a), Dots (Path b)) -> pure (Dots (Path (a ++ b)))
+
+camelCased :: String -> String
+camelCased s = do
+  (prev, cur) <- zip (Nothing:map Just s) (map Just s ++ [Nothing])
+  case (prev, cur) of
+    (Just '_', Just x)
+      | isAlpha x        -> pure (toUpper x)
+    (Just '_', Nothing)  -> pure '_'
+    (Just '_', Just '_') -> pure '_'
+    (_, Just '_')        -> empty
+    (_, Just x)          -> pure x
+    (_, _)               -> empty
+
+typeLikeName :: MonadError CompileError m => String -> m String
+typeLikeName ident@(c:cs)
+  | isUpper c = pure (camelCased ident)
+  | isLower c = pure (camelCased (toUpper c : cs))
+  | '_'  == c = pure (camelCased ('X':ident))
+typeLikeName ident = invalidTypeNameError ident
+
+fieldLikeName :: String -> String
+fieldLikeName ident@(c:_)
+  | isUpper c = let (prefix, suffix) = span isUpper ident
+                in map toLower prefix ++ suffix
+fieldLikeName ident = ident
+
+prefixedEnumFieldName :: String -> String -> String
+prefixedEnumFieldName enumName fieldName = enumName <> fieldName
+
+prefixedConName :: MonadError CompileError m => String -> String -> m String
+prefixedConName msgName conName = (msgName ++) <$> typeLikeName conName
+
+-- TODO: This should be ~:: MessageName -> FieldName -> ...; same elsewhere, the
+-- String types are a bit of a hassle.
+prefixedFieldName :: MonadError CompileError m => String -> String -> m String
+prefixedFieldName msgName fieldName = (fieldLikeName msgName ++) <$> typeLikeName fieldName
+
+dpIdentUnqualName :: MonadError CompileError m => DotProtoIdentifier -> m String
+dpIdentUnqualName (Single name)       = pure name
+dpIdentUnqualName (Dots (Path names)) = pure (last names)
+dpIdentUnqualName (Qualified _ next)  = dpIdentUnqualName next
+dpIdentUnqualName Anonymous           = internalError "dpIdentUnqualName: Anonymous"
+
+dpIdentQualName :: MonadError CompileError m => DotProtoIdentifier -> m String
+dpIdentQualName (Single name)       = pure name
+dpIdentQualName (Dots (Path names)) = pure (intercalate "." names)
+dpIdentQualName (Qualified _ _)     = internalError "dpIdentQualName: Qualified"
```

Comment:
This looks like a bug... shouldn't it be `Qualified q n = dpIdentQualName q <> "." <> dpIdentQualName n`? (ignoring the monad for illustrative purposes)

Reply:
Probably.  I suggest adding a test case for the `.proto` code that this will fail on

Reply:
So the only use of this function is this line in `dotProtoServiceD`

```
     packageName <- dpIdentQualName pkgIdent
```

which makes a string name from the package identifier given as a `DotProtoidentifier`... I guess that the package name can never be `Qualified`, so this code is fine?

We should look into straightening out this `DotProtoIdentifier` name resolution mess, but I'd prefer to do it in a later PR.